### ISSUE TITLE
[Bug] WebsocketNotConnectedException: null

### ISSUE THREAD

**Issue by devoxin** [lavalink-devs/Lavalink#51]
I got this issue upon disconnecting the websocket while a player was playing.
https://i.serux.pro/2b0931ec13.png

**Comment by freyacodes** [comment]
I don't believe that exception has any side effects, but it would be nice to get rid of it

**Comment by devoxin** [comment]
Another stacktrace
```
[00:41:28] [ ERROR] [AudioPlayer] Handler of event com.sedmelluq.discord.lavaplayer.player.event.TrackEndEvent@6015ab12 threw an exception.
org.java_websocket.exceptions.WebsocketNotConnectedException: null
	at org.java_websocket.WebSocketImpl.send(WebSocketImpl.java:608)
	at org.java_websocket.WebSocketImpl.send(WebSocketImpl.java:585)
	at lavalink.server.player.EventEmitter.onTrackEnd(EventEmitter.java:61)
	at com.sedmelluq.discord.lavaplayer.player.event.AudioEventAdapter.onEvent(AudioEventAdapter.java:70)
	at com.sedmelluq.discord.lavaplayer.player.AudioPlayer.dispatchEvent(AudioPlayer.java:295)
	at com.sedmelluq.discord.lavaplayer.player.AudioPlayer.stopWithReason(AudioPlayer.java:139)
	at com.sedmelluq.discord.lavaplayer.player.AudioPlayer.stopTrack(AudioPlayer.java:127)
	at lavalink.server.player.Player.stop(Player.java:100)
	at java.util.concurrent.ConcurrentHashMap$ValuesView.forEach(Unknown Source)
	at lavalink.server.io.SocketContext.shutdown(SocketContext.java:129)
	at lavalink.server.io.SocketServer.close(SocketServer.java:99)
	at lavalink.server.io.SocketServer.onClose(SocketServer.java:90)
	at org.java_websocket.server.WebSocketServer.onWebsocketClose(WebSocketServer.java:577)
	at org.java_websocket.WebSocketImpl.closeConnection(WebSocketImpl.java:504)
	at org.java_websocket.WebSocketImpl.closeConnection(WebSocketImpl.java:526)
	at org.java_websocket.server.WebSocketServer.handleIOException(WebSocketServer.java:521)
	at org.java_websocket.server.WebSocketServer.run(WebSocketServer.java:451)
	at java.lang.Thread.run(Unknown Source)
```

**Comment by schnapster** [comment]
I believe these happen when Lavalink's websocket connection to Discord is closed (by Discord, because for example the bot went offline), while Lavalinks websocket connection to the bot is in the process of timing out (cause the bot went offline, but didn't properly close it, for example).
To prevent these, make sure to properly close the websocket connection from the lavalink client to the lavalink server when shutting down the lavalink client.
The other alternative cause would be network issues impacting both websocket connections.

To sum things up: This is either caused by network issues, or incorrect handling by the client, both events one would be interested in imo.
Given that it does not actually cause any immediate harm it should probably just be caught and logged on warning level, mentioning the possible causes.

**Comment by schnapster** [comment]
I'm going to hold off on implementing what I suggested above.
We (Fred, me) have had some informal talks about implementing Resumes, the outcome of which was "yeah resumes would be good to have", and this case would be one of the things handled by that.

**Comment by ghost** [comment]
I'm getting these exceptions logged when Lavalink is trying to send stats - so every minute. Even though the client is staying connected, with no notice of a disconnect. Sending and receiving events works fine (such as play and trackend). It's even a localhost connection.

```
4|Lavalink | 2018-08-26 20:11:58.754 ERROR 18425 --- [ool-43-thread-1] lavalink.server.io.StatsTask             : Exception while sending stats
4|Lavalink | org.java_websocket.exceptions.WebsocketNotConnectedException: null
4|Lavalink |    at org.java_websocket.WebSocketImpl.send(WebSocketImpl.java:608) ~[Java-WebSocket-1.3.8.jar!/:na]
4|Lavalink |    at org.java_websocket.WebSocketImpl.send(WebSocketImpl.java:585) ~[Java-WebSocket-1.3.8.jar!/:na]
4|Lavalink |    at lavalink.server.io.StatsTask.sendStats(StatsTask.java:119) ~[classes!/:na]
4|Lavalink |    at lavalink.server.io.StatsTask.run(StatsTask.java:53) ~[classes!/:na]
4|Lavalink |    at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:514) [na:na]
4|Lavalink |    at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305) [na:na]
4|Lavalink |    at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305) [na:na]
4|Lavalink |    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135) [na:na]
4|Lavalink |    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [na:na]
4|Lavalink |    at java.base/java.lang.Thread.run(Thread.java:844) [na:na]
```

Also is there a reason why errors get logged to stdout and not stderr?

**Comment by freyacodes** [comment]
>Also is there a reason why errors get logged to stdout and not stderr?

Because logback is not configured to log to stderr by default

**Comment by Diniboy1123** [comment]
I get the same error all the time too.

![image](https://user-images.githubusercontent.com/29148882/45964078-0d0bcc00-c025-11e8-9973-b4bcb21b6ddf.png)

![image](https://user-images.githubusercontent.com/29148882/45964094-18f78e00-c025-11e8-9d66-a8c77ce9d046.png)

Already happened 5700 times in the last 6 days.

What can possibly cause that? A firewall maybe?

**Comment by ghost** [comment]
@Diniboy1123 Unlikely. I have the same problem with localhost connections.

**Comment by Diniboy1123** [comment]
What IP bind do you have now? I am currently using 127.0.0.1 but I can try 0.0.0.0 aswell.

**Comment by ghost** [comment]
127.0.0.1 - for external connections I use a Nginx reverse proxy with TLS.

**Comment by Diniboy1123** [comment]
I also tried to bind it to 0.0.0.0, not helped. And I somehow got an error like this too while I restarted the server:

```
java.util.ConcurrentModificationException: null
    at java.util.HashMap$Values.forEach(HashMap.java:984)
    at lavalink.server.io.StatsTask.sendStats(StatsTask.java:65)
    at lavalink.server.io.StatsTask.run(StatsTask.java:53)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:514)
    at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
    at java.lang.Thread.run(Thread.java:844)
```

**Comment by freyacodes** [comment]
FYI this is not a network issue. The problem is that we sometimes attempt to use a closed socket, whether it has closed unexpectedly or not.

Unrelated errors should be reported separately.

**Comment by Diniboy1123** [comment]
but it happens like every 2 minutes and overall 1800+ issues isn't so good for my Sentry plan. Do you plan to fix it?

**Comment by freyacodes** [comment]
@Diniboy1123 I would prefer if the community did something to fix this, as it isn't really a problem for FredBoat. I might still do it when I have the time.

**Comment by Ahriana** [comment]
i'm also getting this in quite an extreme scale, this often leads to lava imploding and crashing out
![image](https://user-images.githubusercontent.com/28003629/46293831-6d9e8a00-c58c-11e8-8cee-659a4feca83d.png)

not sure how to combat it...

**Comment by freyacodes** [comment]
@RosieCode95 we operate on a much larger scale, and we are pretty much unaffected by that exception. It is likely something else causing your JVM to "implode".

Though I don't have any NPEs tracked in Sentinel. Could you share the one in your screenshot?

**Comment by Ahriana** [comment]
this is the NPE thats crashing the player
```
java.lang.NullPointerException: null
    at lavalink.server.io.SocketServer.close(SocketServer.java:130)
    at lavalink.server.io.SocketServer.onClose(SocketServer.java:122)
    at org.java_websocket.server.WebSocketServer.onWebsocketClose(WebSocketServer.java:595)
    at org.java_websocket.WebSocketImpl.closeConnection(WebSocketImpl.java:504)
    at org.java_websocket.WebSocketImpl.closeConnection(WebSocketImpl.java:515)
    at org.java_websocket.WebSocketImpl.eot(WebSocketImpl.java:563)
    at org.java_websocket.SocketChannelIOHelper.read(SocketChannelIOHelper.java:42)
    at org.java_websocket.server.WebSocketServer.run(WebSocketServer.java:406)
    at java.lang.Thread.run(Thread.java:834)
```

**Comment by freyacodes** [comment]
The WebSocket you are using to manage Lavalink is closing. When this happens it is expected that all music for that connection stops, as there is no mechanism to resume.

**Comment by Ahriana** [comment]
we have been doing our own discord lib etc so been restarting very frequently i **think** this issue is caused by the bot restarting / reconnecting then the player's discord connection timing out. tho its been quite a pain to reproduce.

**Comment by Ahriana** [comment]
the issue is the player totally crashes out, causing the other shards to lose connection etc

**Comment by freyacodes** [comment]
This appears to be unrelated to this issue, please open a new one.

Lavalink Server shouldn't be able to cause you to lose shards, however.

**Comment by Ahriana** [comment]
no problem, i will do that

**Comment by freyacodes** [comment]
#149 should fix this issue. Feel free to [try it out](https://ci.fredboat.com/repository/download/Lavalink_Build/5219:id/Lavalink.jar?guest=1)

You will need to update your config though: https://github.com/Frederikam/Lavalink/commit/68be2ecbe23a29991b7fc321aed185b4bafc193b
